src/Common/StackTrace.cpp

```diff
@@ -496,67 +495,47 @@ static StackTraceCache & cacheInstance()
     return cache;
 }
 
-static std::mutex stacktrace_cache_mutex;
+static DB::SharedMutex stacktrace_cache_mutex;
 
 String toStringCached(const StackTrace::FramePointers & pointers, size_t offset, size_t size)
 {
     const StackTraceRefTriple key{pointers, offset, size};
 
     /// Calculation of stack trace text is extremely slow.
-    /// We use simple cache because otherwise the server could be overloaded by trash queries.
+    /// We use cache because otherwise the server could be overloaded by trash queries.
     /// Note that this cache can grow unconditionally, but practically it should be small.
```

Comment:
Yes, I was also thinking about it. We can lower memory consumption by a factor equal to the stack average depth. And a bit less work for the frames appearing in many stacktraces. Ideally we need to move our cache to `StackTrace::forEachFrame` function. Though, it will require some effort.